crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3298 @@
+                                if self > rhs {
+                                    match rhs.neg_opt() {
+                                        Some(e) => Some(_add(e, self.clone())),
+                                        None => Some(_add(_neg(rhs.clone()), self.clone())),
+                                    }
+                                } else {
+                                    None
+                                }
+                            }
+                        }
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    }
+
+    /// Mul with heuristic optimization
+    fn mul_opt(&self, rhs: &SymbolExpr, recursive: bool) -> Option<SymbolExpr> {
+        if self.is_zero() {
+            Some(self.clone())
+        } else if rhs.is_zero() || self.is_one() {
+            Some(rhs.clone())
+        } else if rhs.is_one() {
+            Some(self.clone())
+        } else if self.is_minus_one() {
+            match rhs.neg_opt() {
+                Some(e) => Some(e),
+                None => Some(_neg(rhs.clone())),
+            }
+        } else if rhs.is_minus_one() {
+            match self.neg_opt() {
+                Some(e) => Some(e),
+                None => Some(_neg(self.clone())),
+            }
+        } else {
+            if let SymbolExpr::Value(_) | SymbolExpr::Symbol(_) = rhs {
+                if let SymbolExpr::Unary { op: _, expr: _ } = self {
+                    return match rhs.mul_opt(self, recursive) {
+                        Some(e) => Some(e),
+                        None => Some(_mul(rhs.clone(), self.clone())),
+                    };
+                }
+            }
+
+            match self {
+                SymbolExpr::Value(e) => e.mul_opt(rhs, recursive),
+                SymbolExpr::Symbol(e) => match rhs {
+                    SymbolExpr::Value(_) => Some(_mul(rhs.clone(), self.clone())),
+                    SymbolExpr::Symbol(r) => {
+                        if r < e {
+                            Some(_mul(rhs.clone(), self.clone()))
+                        } else {
+                            None
+                        }
+                    }
+                    SymbolExpr::Unary {
+                        op: UnaryOp::Neg,
+                        expr,
+                    } => match expr.as_ref() {
+                        SymbolExpr::Value(v) => Some(_mul(SymbolExpr::Value(-v), self.clone())),
+                        SymbolExpr::Symbol(s) => {
+                            if s < e {
+                                Some(_neg(_mul(*expr.clone(), self.clone())))
+                            } else {
+                                Some(_neg(_mul(self.clone(), expr.as_ref().clone())))
+                            }
+                        }
+                        SymbolExpr::Binary {
+                            op: _,
+                            lhs: _,
+                            rhs: _,
```

Comment:
I wonder if you can do this instead?
```suggestion
                            ..
```

Reply:
I did not know `..`, thank you, I replaced for other parts